Code quality and pylint improvements to `A` (#11)
XanaduAI/flamingpy: .github/CHANGELOG.md

> ## Release 0.6.0a3 (development release)

Reviewer: Does the `a3` stand for alpha release number 3? Where's `a1` and `a2`? Is this tag actually necessary? 🤔

Author: Yes, that shows our alpha releases. I believe @ilan-tz, myself, and others agreed it is best to consider the current versions as alpha given there will be new versions comping with new and often breaking structures and features. 

`a1` is released, so all good there. I believe `a2` we decided to not release to PyPI (@ilan-tz knows more details) -- either way, since we are at alpha 3, I suggest we continue the path, conforming to PEP440, to our beta and then removing pre-release tags.

### New features since the last release

### Bug fixes

> ### Housekeeping changes
> 
> * A large number of linting corrections were made to improve the overall pylint report. These were mostly minor but essential modifications including restructuring, re-coding, optimization, updating `.pylintrc`, adding `.coveragerc`. The code quality score is improved to `A` for the released version. Check ["Files changed"](https://github.com/XanaduAI/flamingpy/pull/11/files) for details.
> * `TODO` comments have been removed exclusively from files with low code quality grades. The Dev team has created tickets to be actioned for all removed TODO comments on separate (private) FlamingPy boards.
> * `CONTRIBUTING.md`, `code_of_conduct.md`, and `CHNAGLELOG.md` were imported and modified from the StrawberryFields project. Dev team plans to extend these with customized details in future PRs.
> 

Reviewer: Just suggesting changes to conform to PSS standards. We usually add a link to the PR where the change was added, so that it's easier to track (added this PR to all entries; not sure if that's correct).

```suggestion
### Improvements

* A large number of linting corrections were made to improve the overall pylint report. These were mostly minor but essential modifications including restructuring, re-coding, optimization, updating `.pylintrc`, adding `.coveragerc`. The code quality score is improved to `A` for the released version. Check ["Files changed"](https://github.com/XanaduAI/flamingpy/pull/11/files) for details. [(#11)](https://github.com/XanaduAI/flamingpy/pull/11)

* `TODO` comments have been removed exclusively from files with low code quality grades. The Dev team has created tickets to be actioned for all removed TODO comments on separate (private) FlamingPy boards. [(#11)](https://github.com/XanaduAI/flamingpy/pull/11)

* `CONTRIBUTING.md`, `code_of_conduct.md`, and `CHANGLELOG.md` were imported and modified from the Strawberry Fields project. Dev team plans to extend these with customized details in future PRs. [(#11)](https://github.com/XanaduAI/flamingpy/pull/11)

```

### Documentation changes

### Contributors

> This release contains contributions from (in alphabetical order) ... See full commit details ...

Reviewer: We would also add all contributors here; each time someone new contributes to a specific release, they would be added.
```suggestion
This release contains contributions from (in alphabetical order):

[Nariman Saadatmand](https://github.com/nariman87), [Ilan Tzitrin](https://github.com/ilan-tz)
```

Author: Thanks, for now, I'll prefer to either use my Name with no link to GitHub or only my handle everywhere, please. I'll remove the link for myself, happy with the rest.

---

## Release 0.4.9a1 (current release)

> <h3>Housekeeping changes since the last release</h3>

Reviewer: ```suggestion
### Improvements
```

Author: I suggest it's useful to always say "since the last release" for the top one (as in changing "### Improvements since the last release" or if we have another title on top adding "since the last release") -- this way it is clear what we are comparing against even if there are no new features.

* Minor updates to `README.md` following the start of the public project. This included:
   * Relative paths cannot be used in README.md logos and were replaced with Github-hosted links.
* C++ imports are now placed within `try` blocks to avoid interrupting non-compiled installations, such as the ones currently used by readthedocs.
* Code coverage (quality) score was improved to a more acceptable `B-` level.

> <h3>Bug fixes</h3>

Reviewer: ```suggestion
### Bug fixes
```

* Fixed a bug in [pull_request_template.md](https://github.com/XanaduAI/flamingpy/pull/2/commits/e30f2cb65daffece08b193ffc4b8fe7a8d90b90e). The template was not loading properly due to a whitespace problem.
* Fixed a bug in [`simulations.py`](flamingpy/simulations.py) and related examples. See [here](https://github.com/XanaduAI/flamingpy/commit/771b0e66e5471c3696ac2e779a2df1cc49e5e684) for details.

> <h3>Documentation changes</h3>

Reviewer: ```suggestion
### Documentation changes
```

* Making Documentation more usable and consistent with other Xanadu projects:
   * API details and inheritance diagrams should be now correctly displayed.
   * "Edit on Github" links were fixed
   * The general style and section structures made more consistent with the company requirements and other packages such as StrawberryFields.
   * Fixed the documentation links in `README.md`
   * Minor updates to `doc/conf.py`, `doc/dev_requirements.txt`, and `doc/Makefile`. 

> <h3>Contributors</h3>
> 
> This release contains contributions from (in alphabetical order) @nariman87 and Ilan Tzitrin. See full commit details https://github.com/XanaduAI/flamingpy/compare/v0.4.6a1...v0.4.9a1.

Reviewer: ```suggestion
### Contributors

This release contains contributions from (in alphabetical order):
[Nariman Saadatmand](https://github.com/nariman87), [Ilan Tzitrin](https://github.com/ilan-tz)

See full commit details https://github.com/XanaduAI/flamingpy/compare/v0.4.6a1...v0.4.9a1.
```

Author: Same as my comment above.

## Release 0.4.6a1

> <h3>New features since the last private release</h3>

Reviewer: ```suggestion
### New features since the last private release
```

* The first Cython function for Monte Carlo sampling, mostly to provide cythonization samples and testbeds, has been added. See [`cpp_mc_loop.pyx`](flamingpy/cpp/cpp_mc_loop.pyx) and [`simulations.py`](flamingpy/benchmarks/simulations.py) for detailes.

> <h3>Housekeeping changes</h3>

Reviewer: ```suggestion
### Improvements
```

* More options for Installation from Source: 
`setup.py` was updated to provide a no-compilation option for only installing (purely) Python libraries and separate options to compile `cmake` and `cython`-based codes. See the new [README.md](https://github.com/XanaduAI/ft-stack/blob/mc-cpp/README.md) for details.

> <h3>Contributors</h3>

Reviewer: ```suggestion
### Contributors
```

> This release contains contributions from (in alphabetical order) @nariman87 and Ilan Tzitrin.

Reviewer: ```suggestion
This release contains contributions from (in alphabetical order):
[Nariman Saadatmand](https://github.com/nariman87), [Ilan Tzitrin](https://github.com/ilan-tz)
```

Author: Same as my comment above.

* `TODO` comments have been removed exclusively from files with low code quality grades. The Dev team has created tickets to be actioned for all removed TODO comments on separate (private) FlamingPy boards. [(#11)](https://github.com/XanaduAI/flamingpy/pull/11)

* `CONTRIBUTING.md`, `code_of_conduct.md`, and `CHANGLELOG.md` were imported and modified from the Strawberry Fields project. Dev team plans to extend these with customized details in future PRs. [(#11)](https://github.com/XanaduAI/flamingpy/pull/11)

### Documentation changes

### Contributors

This release contains contributions from (in alphabetical order):

Nariman Saadatmand, [Ilan Tzitrin](https://github.com/ilan-tz)

See full commit details ...

---

## Release 0.4.9a1 (current release)

> <h3>Housekeeping changes since the last release</h3>

Reviewer: ```suggestion
### Housekeeping changes since the last release
```

Author: Has been fixed separately.

> * Minor updates to `README.md` following the start of the public project. This included:
>    * Relative paths cannot be used in README.md logos and were replaced with Github-hosted links. [(#5)](https://github.com/XanaduAI/flamingpy/pull/5)
> * C++ imports are now placed within `try` blocks to avoid interrupting non-compiled installations, such as the ones currently used by readthedocs. [(#5)](https://github.com/XanaduAI/flamingpy/pull/5)
> * Code coverage (quality) score was improved to a more acceptable `B-` level. [(#5)](https://github.com/XanaduAI/flamingpy/pull/5)

Reviewer: ```suggestion
* Relative paths cannot be used in README.md logos and were replaced with Github-hosted links. [(#5)](https://github.com/XanaduAI/flamingpy/pull/5)

* C++ imports are now placed within `try` blocks to avoid interrupting non-compiled installations, such as the ones currently used by readthedocs. [(#5)](https://github.com/XanaduAI/flamingpy/pull/5)

* Code coverage (quality) score was improved to a more acceptable `B-` level. [(#5)](https://github.com/XanaduAI/flamingpy/pull/5)
```

### Bug fixes

> * Fixed a bug in [pull_request_template.md](https://github.com/XanaduAI/flamingpy/pull/2/commits/e30f2cb65daffece08b193ffc4b8fe7a8d90b90e). The template was not loading properly due to a whitespace problem. [(#2)](https://github.com/XanaduAI/flamingpy/pull/2)
> * Fixed a bug in [`simulations.py`](flamingpy/simulations.py) and related examples. See [here](https://github.com/XanaduAI/flamingpy/commit/771b0e66e5471c3696ac2e779a2df1cc49e5e684) for commit details. [(#6)](https://github.com/XanaduAI/flamingpy/pull/6)

Reviewer: ```suggestion
* Fixed a bug in [pull_request_template.md](https://github.com/XanaduAI/flamingpy/pull/2/commits/e30f2cb65daffece08b193ffc4b8fe7a8d90b90e). The template was not loading properly due to a whitespace problem. [(#2)](https://github.com/XanaduAI/flamingpy/pull/2)

* Fixed a bug in [`simulations.py`](flamingpy/simulations.py) and related examples. See [here](https://github.com/XanaduAI/flamingpy/commit/771b0e66e5471c3696ac2e779a2df1cc49e5e684) for commit details. [(#6)](https://github.com/XanaduAI/flamingpy/pull/6)
```

   * Fixed the documentation links in `README.md`
   * Minor updates to `doc/conf.py`, `doc/dev_requirements.txt`, and `doc/Makefile`. 

### Contributors

This release contains contributions from (in alphabetical order):

Nariman Saadatmand, [Ilan Tzitrin](https://github.com/ilan-tz)

See full commit details https://github.com/XanaduAI/flamingpy/compare/v0.4.6a1...v0.4.9a1

## Release 0.4.6a1

### New features since the last private release

* The first Cython function for Monte Carlo sampling, mostly to provide cythonization samples and testbeds, has been added. See [`cpp_mc_loop.pyx`](flamingpy/cpp/cpp_mc_loop.pyx) and [`simulations.py`](flamingpy/benchmarks/simulations.py) for detailes. 

### Improvements

* More options for Installation from Source: 
`setup.py` was updated to provide a no-compilation option for only installing (purely) Python libraries and separate options to compile `cmake` and `cython`-based codes. See the new [README.md](https://github.com/XanaduAI/ft-stack/blob/mc-cpp/README.md) for details.

### Contributors

This release contains contributions from (in alphabetical order):

> Nariman Saadatmand, [Ilan Tzitrin](https://github.com/ilan-tz)

Reviewer 2: Shall we add @sduquemesa here and hyperlink @nariman87 ?

Author: Feel free to do so, @ilan-tz. But, not essential, because one purpose of the Special Release branch and upcoming PRs is to update CHANGELOG.md and make sure all contributors are added. So, we will do this anyway in near future.
